dv/verilator/simutil_verilator/cpp/verilated_toplevel.h

```diff
@@ -97,27 +113,18 @@ class VerilatedToplevel {
   virtual void trace(VerilatedTracer &tfp, int levels, int options) = 0;
 };
 
-#define STR(s) #s
-
-#if VM_TRACE == 1
-#  define VERILATED_TOPLEVEL_TRACE_CALL(topname) \
-    V##topname::trace(static_cast<VM_TRACE_CLASS_NAME *>(tfp), levels, options);
-#else
-#  define VERILATED_TOPLEVEL_TRACE_CALL(topname) \
-    assert(0 && "Tracing not enabled.");
-#endif
-
-#define VERILATED_TOPLEVEL(topname)                                 \
-  class topname : public V##topname, public VerilatedToplevel {     \
-   public:                                                          \
-    topname(const char *name = "TOP")                               \
-        : V##topname(name), VerilatedToplevel() {}                  \
-    const char *name() const { return STR(topname); }               \
-    void eval() { V##topname::eval(); }                             \
-    void final() { V##topname::final(); }                           \
-    void trace(VerilatedTracer &tfp, int levels, int options = 0) { \
-      VERILATED_TOPLEVEL_TRACE_CALL(topname)                        \
-    }                                                               \
-  };
+class TOPLEVEL_NAME : public VERILATED_TOPLEVEL_NAME, public VerilatedToplevel {
+ public:
+  TOPLEVEL_NAME(const char *name = "TOP")
+      : VERILATED_TOPLEVEL_NAME(name), VerilatedToplevel() {}
+  const char *name() const { return STR(TOPLEVEL_NAME); }
+  void eval() { VERILATED_TOPLEVEL_NAME::eval(); }
+  void final() { VERILATED_TOPLEVEL_NAME::final(); }
+  void trace(VerilatedTracer &tfp, int levels, int options = 0) {
+    assert(VM_TRACE && "Tracing not enabled.");
+    VERILATED_TOPLEVEL_NAME::trace(static_cast<VM_TRACE_CLASS_NAME *>(tfp),
```

Comment:
I thought the point in the `ifdef` was to cut code size by not even including this call, if `VM_TRACE` was turned off.

Reply:
We need to keep the `trace()` to fully implement `VerilatedToplevel`, the `#ifdef` code was probably just a leftover from when it wasn't obvious that `VM_TRACE` is actually an integer.

Reply:
After the discussion, it was obvious that we need the `#if` to capture the case where `VERILATED_TOPLEVEL_NAME::trace()` doesn't exist. Added it back.